src/sync.cpp

```diff
@@ -73,7 +73,11 @@ struct LockData {
     LockOrders lockorders;
     InvLockOrders invlockorders;
     std::mutex dd_mutex;
-} static lockdata;
+};
+LockData& get_lockdata() {
+    static LockData* lockdata = new LockData();
```

Comment:
Can you use a `std::unique_ptr<LockData>` here instead? This code will technically cause a memory leak (as the `LockData` won't get cleaned up at shutdown).

Reply:
Hm? Why would it cause a leak? All memory used by a process is always released upon shutdown.

Reply:
@kallewoof Yes, that's why I say technically. We generally ain to follow that rule though, as it means leak detection systems like valgrind don't report spurious leaks.

Reply:
I thought valgrind was intelligent about static vars inside functions, and always considered this type of expression to be acceptable. I guess a unique pointer is less dirty though, so concept ACK on the change.

Reply:
@kallewoof Good point, I hadn't considered that actually. Some googling indicates that in cases like this, the memory would indeed be counted as still reachable rather than lost.

Still, using a unique_ptr here comes at no cost, and is more obviously correct.

Reply:
The consensus must be that we should aim for zero bytes "still reachable" when exiting, right? :-)

Reply:
Is this thread safe?

Reply:
Yes, since C++11 the language guarantees that the static initializer is only run once. If it is invoked from multiple threads simultaneously, only the first will complete, and the others will wait for the first to run and use the same result.

Reply:
Could just be `static LockData lockdata;` right?